### PR TITLE
EOS-25627: Handle parallel multipart POST complete and Put part operations

### PR DESCRIPTION
**Pull request by DPG17**
Signed-off-by: Dattaprasad Govekar <dattaprasad.govekar@seagate.com>

# Problem Statement
-  Parallel operations lead to object leak

# Design
-  Handling object leak due to parallel S3 operations (POST complete, multipart PUT, simple PUT)

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by SeagateChaDeepak — APPROVED** [review]
+1 LGTM

**Review by t7ko-seagate — CHANGES_REQUESTED** [review]

File: s3backgrounddelete/s3backgrounddelete/object_recovery_validator.py [hunk 1]
```diff
@@ -428,6 +428,44 @@ def version_entry_cb(self, versionEntry, current_oid, timeVersionEntry):
...
+            self.object_leak_info["ext_version_id"] + "|" + "P"
+
+        parent_oid = NULL_OBJ_OID
+        if "parent_oid" in self.object_leak_info:
+            parent_oid = self.object_leak_info["parent_oid"]
+
+        if (parent_oid == NULL_OBJ_OID and part_no != 0):
+            # This is parent multipart object. Process extended entries
+            for part in range(1, part_no + 1):
+                extended_key = extended_key_prefix + str(part) + "|" + "F1"
```
[the remark this answers is not in the document]
Comment (DPG17) [reply s3backgrounddelete/s3backgrounddelete/object_recovery_validator.py:457]: S3server, as of now, does not support fragmentation within simple object (which is S3 fault tolerance feature's capability).
Once we have such feature in S3, we would have appropriate changes here.

File: s3backgrounddelete/s3backgrounddelete/object_recovery_validator.py [hunk 2]
```diff
@@ -428,6 +428,44 @@ def version_entry_cb(self, versionEntry, current_oid, timeVersionEntry):
...
+        bStatus = False
+        ext_motr_oid_key = "OID"
+        ext_layout_id_key = "layout-id"
+        ext_pvid_key = "PVID"
+        ext_api_prefix = "EXTENDED MD LIST DEL"
+        if (object_extended_list_index is None):
+            return bStatus
+
+        if "part" in self.object_leak_info:
+            part_no = self.object_leak_info["part"]
```
[the remark this answers is not in the document]
Comment (DPG17) [reply s3backgrounddelete/s3backgrounddelete/object_recovery_validator.py:444]: ok

**Review by t7ko-seagate — APPROVED** [review]

**Comment by t7ko-seagate** [comment]
Merging as per the team agreement that Codacy warnings can be ignored.

**Pull request merged**